File: backend/app/services/workout/cycling_power.py
from __future__ import annotations


def estimate_cycling_power_watts(
    *,
    distance_miles: float | None,
    duration_seconds: int | float | None,
    rider_weight_lbs: float | None,
    elevation_gain_ft: float | None = None,
    indoor: bool = False,
) -> int | None:
    try:
        distance = float(distance_miles or 0)
        duration = float(duration_seconds or 0)
        weight_lbs = float(rider_weight_lbs or 0)
        gain_ft = float(elevation_gain_ft or 0)
    except (TypeError, ValueError):
        return None
    if distance <= 0 or duration <= 0 or weight_lbs <= 0:
        return None

    mph = distance / (duration / 3600)
    if mph < 3 or mph > 45:
        return None

    distance_m = distance * 1609.344
    speed_mps = distance_m / duration
    rider_kg = weight_lbs * 0.45359237
    bike_kg = 10.0 if indoor else 12.0
    total_mass_kg = rider_kg + bike_kg
    gain_m = max(0.0, gain_ft * 0.3048)
    grade = min(0.18, gain_m / max(1.0, distance_m))

    gravity = 9.80665
    rolling_resistance = 0.004 if indoor else 0.005
    air_density = 1.225
    drag_area = 0.30 if indoor else 0.40
    drivetrain_efficiency = 0.95

    rolling_watts = total_mass_kg * gravity * rolling_resistance * speed_mps
    climbing_watts = total_mass_kg * gravity * grade * speed_mps
    aero_watts = 0.5 * air_density * drag_area * (speed_mps ** 3)
    watts = (rolling_watts + climbing_watts + aero_watts) / drivetrain_efficiency
    if not (25 <= watts <= 650):
        return None
    return int(round(watts))

File: src/utils/cyclingPower.ts
export function estimateCyclingPowerWatts({
  distanceMiles,
  durationSeconds,
  riderWeightLbs,
  elevationGainFt = 0,
  indoor = false,
}: {
  distanceMiles?: number | null;
  durationSeconds?: number | null;
  riderWeightLbs?: number | null;
  elevationGainFt?: number | null;
  indoor?: boolean;
}): number | null {
  const distance = Number(distanceMiles ?? 0);
  const duration = Number(durationSeconds ?? 0);
  const weightLbs = Number(riderWeightLbs ?? 0);
  const gainFt = Number(elevationGainFt ?? 0);
  if (!Number.isFinite(distance) || !Number.isFinite(duration) || !Number.isFinite(weightLbs)) return null;
  if (distance <= 0 || duration <= 0 || weightLbs <= 0) return null;

  const mph = distance / (duration / 3600);
  if (mph < 3 || mph > 45) return null;

  const distanceM = distance * 1609.344;
  const speedMps = distanceM / duration;
  const riderKg = weightLbs * 0.45359237;
  const bikeKg = indoor ? 10.0 : 12.0;
  const totalMassKg = riderKg + bikeKg;
  const gainM = Math.max(0, Number.isFinite(gainFt) ? gainFt * 0.3048 : 0);
  const grade = Math.min(0.18, gainM / Math.max(1, distanceM));

  const gravity = 9.80665;
  const rollingResistance = indoor ? 0.004 : 0.005;
  const airDensity = 1.225;
  const dragArea = indoor ? 0.30 : 0.40;
  const drivetrainEfficiency = 0.95;

  const rollingWatts = totalMassKg * gravity * rollingResistance * speedMps;
  const climbingWatts = totalMassKg * gravity * grade * speedMps;
  const aeroWatts = 0.5 * airDensity * dragArea * (speedMps ** 3);
  const watts = (rollingWatts + climbingWatts + aeroWatts) / drivetrainEfficiency;
  if (!Number.isFinite(watts) || watts < 25 || watts > 650) return null;
  return Math.round(watts);
}
